pkg/scheduler/plugins/elasticquota/plugin.go

```diff
@@ -0,0 +1,560 @@
+	// PreFilter, refresh elasticQuotaInfos' runtime/used according to groupQuotaManager, then write snapshot.
+	// Reserve/ Unreserve/ OnPodAdd/ OnPodUpdate/ OnPodDelete, update the corresponding SimpleElasticQuotaInfo's pods.
+	elasticQuotaInfos SimpleElasticQuotaInfos
+}
+
+// SnapshotState stores the snapshot of elasticQuotaInfos.
+type SnapshotState struct {
+	elasticQuotaInfos SimpleElasticQuotaInfos
+}
+
+// Clone the ElasticQuotaSnapshot state.
+func (s *SnapshotState) Clone() framework.StateData {
+	return &SnapshotState{
+		elasticQuotaInfos: s.elasticQuotaInfos.clone(),
+	}
+}
+
+var (
+	_ framework.PreFilterPlugin  = &Plugin{}
+	_ framework.PostFilterPlugin = &Plugin{}
+	_ framework.ReservePlugin    = &Plugin{}
+)
+
+func New(args runtime.Object, handle framework.Handle) (framework.Plugin, error) {
+	pluginArgs, ok := args.(*config.ElasticQuotaArgs)
+	if !ok {
+		return nil, fmt.Errorf("want args to be of type GangSchedulingArgs, got %T", args)
+	}
+	if err := validation.ValidateElasticQuotaArgs(pluginArgs); err != nil {
+		return nil, err
+	}
+
+	client, ok := handle.(versioned.Interface)
+	if !ok {
+		kubeConfig := *handle.KubeConfig()
+		kubeConfig.ContentType = runtime.ContentTypeJSON
+		kubeConfig.AcceptContentTypes = runtime.ContentTypeJSON
+		client = versioned.NewForConfigOrDie(&kubeConfig)
+	}
+	scheSharedInformerFactory := externalversions.NewSharedInformerFactory(client, 0)
+	elasticQuotaInformer := scheSharedInformerFactory.Scheduling().V1alpha1().ElasticQuotas()
+
+	elasticQuota := &Plugin{
+		handle:            handle,
+		pluginArgs:        pluginArgs,
+		podLister:         handle.SharedInformerFactory().Core().V1().Pods().Lister(),
+		quotaLister:       elasticQuotaInformer.Lister(),
+		pdbLister:         getPDBLister(handle.SharedInformerFactory()),
+		nodeLister:        handle.SharedInformerFactory().Core().V1().Nodes().Lister(),
+		groupQuotaManager: core.NewGroupQuotaManager(pluginArgs.SystemQuotaGroupMax, pluginArgs.DefaultQuotaGroupMax),
+		nodeResourceMap:   make(map[string]struct{}),
+		elasticQuotaInfos: NewSimpleElasticQuotaInfos(),
+	}
+	elasticQuota.groupQuotaManager.SetScaleMinQuotaEnabled(true)
+	elasticQuota.quotaOverUsedRevokeController = NewQuotaOverUsedRevokeController(
+		*elasticQuota.pluginArgs.ContinueOverUseCountTriggerEvict, elasticQuota.groupQuotaManager)
+
+	nodeInformer := handle.SharedInformerFactory().Core().V1().Nodes().Informer()
+	nodeInformer.AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    elasticQuota.OnNodeAdd,
+			UpdateFunc: elasticQuota.OnNodeUpdate,
+			DeleteFunc: elasticQuota.OnNodeDelete,
+		})
+
+	elasticQuotaInformer.Informer().AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    elasticQuota.OnQuotaAdd,
+			UpdateFunc: elasticQuota.OnQuotaUpdate,
+			DeleteFunc: elasticQuota.OnQuotaDelete,
+		})
+
+	elasticQuota.elasticQuotaController = NewElasticQuotaController(kubernetes.NewForConfigOrDie(handle.KubeConfig()),
```

Comment:
plugin noneed keep a elasticQuotaController pointer,  put elasticQuotaController init logic to controller itself

Reply:
fixed